fix: refresh the rollup index as part of the rollup indexer actions

salvatore-campagna opened the pull request:
Before reading documents from the rollup index we need the index
to be refreshed to avoid the search API not hitting the expected results.
Here, instead of relying on the test to request a refresh, we let the
rollup job do the refresh of the rollup index as part of the `onFinish`
callback.

Resolves #81983
Resolves #53412

salvatore-campagna committed:
fix: move refresh just before the match_all query

Using the command line provided in #81983 I could not reproduce the issue.
Anyway, changing the timeout from 5 seconds to just 1 second I could
reproduce the issue. At the end, keeping a timeout of 1 second I could fix
the issue moving the refresh operation just before running the match_all
query. My understanding is that the timeout is used to give the rollup job
enough time to complete the rollup operation. Anyway, other than that, after
the rollup job has written data into the rollup index we need to refresh it
in order for the subsequent search operation to hit the expected rolledup
document.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

salvatore-campagna commented:
> 

I will see in the API if we have such an option.

salvatore-campagna commented:
I think this patch might address also https://github.com/elastic/elasticsearch/issues/76231

csoulios reviewed (commented):
Rollup v1 indeed does not refresh the index after adding the new rollup documents. 

Moving the refresh action before the search would probably fix this test. However, I was thinking that it may be preferrable to add a refresh at the end of the rollup job.

csoulios commented:
> I genuinely unsure what's going on here. I wish we didn't have to delay some number of seconds like this. If the action is async we should be returning a task_id that we can wait_for_completion on. Is that a thing here?

Rollup v1 works by scheduling a cron job (persistent task) and it's not an asyc action. So, the delay is added to make sure that the job has started and completed after job `start_job` has been called and the rollup job has completed.

salvatore-campagna committed:
fix: refresh the rollup index when rollup is done

salvatore-campagna committed:
Revert "fix: move refresh just before the match_all query"

Instead of refreshing from the yaml test we refresh once the
indexing operation is complete in the callback.

This reverts commit 7968c70df478366c6a9326484c35678c373a9b4a.

elasticmachine committed:
Merge branch 'master' into fix/81983-security-tests-rollup

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/job/RollupIndexer.java:
```diff
@@ -115,6 +126,12 @@ protected void onStart(long now, ActionListener<Boolean> listener) {
+    @Override
+    protected void onFinish(ActionListener<Void> listener) {
+        final RefreshRequest refreshRequest = new RefreshRequest(job.getConfig().getRollupIndex());
+        client.admin().indices().refresh(refreshRequest).actionGet();
```
What happens if the call to `refresh` fails? `actionGet()` may throw exceptions here?

Also, I think the call to `refresh` sould be async by calling method

```java
client.admin().indices().refresh(RefreshRequest request, ActionListener<RefreshResponse> listener)
```

The big question is what happens if the call to `refresh` fails. 
Since we got there, the index has been created successfully and we should not fail the whole job.

salvatore-campagna replied:
I guess in both cases, successful or not we can just log something.

csoulios commented on x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/rollup/job/RollupIndexerIndexingTests.java:
```diff
@@ -705,10 +707,12 @@ private void executeTestCase(
         IndexSearcher searcher = new IndexSearcher(reader);
         String dateHistoField = config.getGroupConfig().getDateHistogram().getField();
         final ThreadPool threadPool = new TestThreadPool(getTestName());
+        final Client client = Mockito.mock(Client.class);
```
Perhaps you would like to use `NoOpClient` here and everywhere else. Although Mockito should do its job fine here, we generally prefer this as to mocking the class directly.

csoulios reviewed (commented):
Generally the change looks good. I left some small comments

salvatore-campagna committed:
fix: log refresh failure

salvatore-campagna committed:
use NoOpClient in place of Mockito mock

salvatore-campagna committed:
fix: remove redundant 'final' modifier

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/81983-security-tests-rollup

csoulios reviewed (commented):
Maybe it's worth trying implementing this logic in `ClientRollupPageManager`

csoulios commented:
Also, this PR fixes https://github.com/elastic/elasticsearch/issues/53412

salvatore-campagna committed:
fix: move 'onFinish' to subclass

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/81983-security-tests-rollup

salvatore-campagna committed:
fix: just mock the search action call

The client is used to call the refresh action too. As a result, using
'any()' as the action would match both the search and refresh actions
while we just need to handle the search action.

csoulios reviewed (commented):
Left some more comments. Basically, now that refresh happens in `ClientRollupPageManager`, we don't need the `client` dependency in the `RollupIndexer` class and all its subclasses.

salvatore-campagna committed:
fix: remove unnecessary client dependency

salvatore-campagna committed:
fix: call onResponse from inside and change logging level

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix/81983-security-tests-rollup

salvatore-campagna committed:
fix: move back onResponse call

csoulios reviewed (approved):
Generally, change looks good. 

I have only left a comment that the `client` arg should be removed from the `SyncRollupIndexer`. I don't think `Client` is used in this class. After changing this, no other review is required.

Thank you for fixing this bug.

salvatore-campagna committed:
fix: remove client dependency

salvatore-campagna committed:
fix: move listener callback into the refresh listener

We also need to mock the refresh response for the client
to actually propagate the result of the refresh call.
